src/backend/access/aocs/aocsam.c

```diff
@@ -1199,21 +1202,25 @@ aocs_gettuple(AOCSScanDesc scan, int64 targrow, TupleTableSlot *slot)
 			}
 		}
 
-		if (ds->blockRowCount <= 0)
+		if (scan->segrowsprocessed == 0)
 			; /* haven't read block */
 		else
 		{
 			/* block was read */
 			rowcount = aocs_block_remaining_rows(ds);
 			Assert(rowcount >= 0);
 
-			if (startrow + rowcount - 1 >= targrow)
+			if (isanchor ?
```

Comment:
I think that the complex conditions here and in the function parameters do complicate the overall function readability. I would suggest to rewrite a bit to improve readability by using additional variables, for ex:
```diff
diff --git a/src/backend/access/aocs/aocsam.c b/src/backend/access/aocs/aocsam.c
index db884230ef..17a121a7ee 100644
--- a/src/backend/access/aocs/aocsam.c
+++ b/src/backend/access/aocs/aocsam.c
@@ -1182,6 +1182,7 @@ aocs_gettuple(AOCSScanDesc scan, int64 targrow, TupleTableSlot *slot)
 		bool chkvisimap = true;
 		bool isanchor = attno == scan->columnScanInfo.proj_atts[ANCHOR_COL_IN_PROJ];
 
+
 		/*
 		 * After scanning the anchor column, we check if the value is 
 		 * missing in this column, if so, we skip the scanning.
@@ -1209,16 +1210,18 @@ aocs_gettuple(AOCSScanDesc scan, int64 targrow, TupleTableSlot *slot)
 			/* block was read */
 			rowcount = aocs_block_remaining_rows(ds);
 			Assert(rowcount >= 0);
+			int64 current_start_row = isanchor ? startrow : ds->blockFirstRowNum;
+			int64 current_remaining_rows = isanchor ? rowcount : ds->blockRowCount;
+			int64 current_target_row = isanchor ? targrow : phyrow;
 
-			if (isanchor ?
-				startrow + rowcount - 1 >= targrow :
-				ds->blockFirstRowNum + ds->blockRowCount - 1 >= phyrow)
+			if (current_start_row + current_remaining_rows - 1 >= current_target_row)
 			{
+				current_start_row += (isanchor ? ds->blockRowsProcessed : 0);
 				/* read the value, visimap check only needed for the anchor column */
 				phyrow = aocs_gettuple_column(scan,
 											  attno,
-											  isanchor ? startrow : ds->blockFirstRowNum + ds->blockRowsProcessed,
-											  isanchor ? targrow : phyrow,
+											  current_start_row,
+											  current_target_row,
 											  isanchor ? &chkvisimap : NULL,
 											  slot);
 				if (!chkvisimap)
@@ -1242,12 +1245,13 @@ aocs_gettuple(AOCSScanDesc scan, int64 targrow, TupleTableSlot *slot)
 				rowcount = ds->blockRowCount;
 				Assert(rowcount > 0);
 
+				int64 current_start_row = isanchor ? startrow : ds->blockFirstRowNum;
+				int64 current_target_row = isanchor ? targrow : phyrow;
+
 				/* new block, reset blockRowsProcessed */
 				ds->blockRowsProcessed = 0;
 
-				if (isanchor ?
-					startrow + rowcount - 1 >= targrow :
-					ds->blockFirstRowNum + ds->blockRowCount - 1 >= phyrow)
+				if (current_start_row + rowcount - 1 >= current_target_row)
 				{
 					int64 blocksRead;
 
@@ -1260,11 +1264,13 @@ aocs_gettuple(AOCSScanDesc scan, int64 targrow, TupleTableSlot *slot)
 					pgstat_count_buffer_read_ao(scan->rs_base.rs_rd,
 												blocksRead);
 
+					current_start_row += (isanchor ? ds->blockRowsProcessed : 0);
+
 					/* read the value, visimap check only needed for the anchor column */
 					phyrow = aocs_gettuple_column(scan,
 												  attno,
-												  isanchor ? startrow : ds->blockFirstRowNum + ds->blockRowsProcessed,
-												  isanchor ? targrow : phyrow,
+												  current_start_row,
+												  current_target_row,
 												  isanchor ? &chkvisimap : NULL,
 												  slot);
 					if (!chkvisimap)
```